apollo-router/src/uplink/entitlement.rs

```diff
@@ -1,32 +1,181 @@
 // With regards to ELv2 licensing, this entire file is license key functionality
-
+use std::fmt::Display;
+use std::fmt::Formatter;
 use std::str::FromStr;
 use std::time::Duration;
+use std::time::SystemTime;
 
+use buildstructor::Builder;
 use displaydoc::Display;
 use futures::Stream;
+use itertools::Itertools;
+use jsonwebtoken::decode;
+use jsonwebtoken::jwk::JwkSet;
+use jsonwebtoken::DecodingKey;
+use jsonwebtoken::Validation;
+use once_cell::sync::OnceCell;
+use regex::Regex;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json::Value;
 use thiserror::Error;
 use url::Url;
 
+use crate::spec::Schema;
+use crate::Configuration;
+
+static JWKS: OnceCell<JwkSet> = OnceCell::new();
+
 #[derive(Error, Display, Debug)]
 pub enum Error {
-    /// invalid entitlement
-    InvalidEntitlement,
+    /// invalid entitlement: {0}
+    InvalidEntitlement(jsonwebtoken::errors::Error),
+
+    /// entitlement violations: {0}
+    EntitlementViolations(EntitlementReport),
+}
+
+#[derive(Eq, PartialEq)]
+enum RouterState {
+    Startup,
+    Running,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum EntitlementState {
+    Oss,
+    Entitled,
+    Warning,
+    Halt,
+}
+
+#[derive(Eq, PartialEq)]
+enum Action {
+    PreventStartup,
+    PreventReload,
+    Warn,
+    Halt,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+enum Audience {
+    SelfHosted,
+    Cloud,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(untagged)]
+enum OneOrMany<T> {
+    One(T),
+    Many(Vec<T>),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+struct Claims {
+    iss: String,
+    sub: String,
+    aud: OneOrMany<Audience>,
+    #[serde(with = "serde_millis", rename = "warnAt")]
+    warn_at: SystemTime,
+    #[serde(with = "serde_millis", rename = "haltAt")]
+    halt_at: SystemTime,
+}
+
+impl Claims {
+    fn entitlement_state(&self) -> EntitlementState {
+        let now = SystemTime::now();
+        if self.halt_at < now {
+            EntitlementState::Halt
+        } else if self.warn_at < now {
+            EntitlementState::Warning
+        } else {
+            EntitlementState::Entitled
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EntitlementReport {
+    entitlement_state: EntitlementState,
+    configuration_violations: Vec<ConfigurationRestriction>,
+}
+
+impl Display for EntitlementReport {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let formatted_violations = self
+            .configuration_violations
+            .iter()
+            .map(|v| format!("* {} at {}", v.name, v.path))
+            .join("\n");
+
+        write!(
+            f,
+            "TODO PREAMBLE\n\nViolations:\n{}\n\nTODO POSTAMBLE",
+            formatted_violations
+        )
+    }
+}
+
+impl EntitlementReport {
+    fn action(&self, state: RouterState) -> Action {
+        // A state of None indicates an OSS user
```

Comment:
what is this `None` state here? It looks like it's never represented

Reply:
Ahh, this is a leftover comment from a previous version. It got replaced with the Oss in the enum. I'll remove the comment.

Reply:
[e31befc](https://github.com/apollographql/router/pull/2521/commits/e31befc6fdfaedd93a774ea20a21c7f14f60d98b)